src/commands/up.rs

```diff
@@ -646,6 +668,76 @@ fn trigger_command_for_resolved_app_source(
         .collect()
 }
 
+// Scrubs the locked app to only contain the given list of components
+// Introspects the LockedApp to find and selectively retain the triggers that correspond to those components
+fn retain_components(locked_app: &mut LockedApp, component_trigger_id_map: ComponentTriggerMap) {
+    let components = component_trigger_id_map.keys().collect::<HashSet<_>>();
+    let triggers = component_trigger_id_map.values().collect::<HashSet<_>>();
+    locked_app.components.retain(|c| components.contains(&c.id));
+    locked_app.triggers.retain(|t| triggers.contains(&t.id));
+}
+
+// Given a list of components and App, returns a map of component ID to trigger ID.
+// Also validates that the components exist in the app and have valid service chaining.
+fn component_trigger_map(
+    app: &spin_app::App,
+    components: &[String],
+) -> Result<ComponentTriggerMap> {
+    let component_triggers: Vec<(AppComponent<'_>, String)> = app
+        .triggers()
+        .filter_map(|t| match t.component() {
+            Ok(comp) => {
+                if components.contains(&comp.id().to_string()) {
+                    Some((comp, t.id().to_string()))
+                } else {
+                    None
+                }
+            }
+            Err(_) => None,
+        })
+        .collect();
+
+    // Allow only service chaining among retained components
+    // All wildcard service chaining is disallowed
+    for (c, _) in &component_triggers {
```

Comment:
:shrug: 
```suggestion
    for c in component_triggers.keys() {
```

Reply:
At this point it is a vector not a HashMap. I agree that this looks a little wonky, but we need the full `AppComponent` object at this point so it is too early to convert it to a `ComponentTriggerMap`.

Reply:
yeah nbd